Some more math support

andswitch opened the pull request:
Added Latex math and Gnuplot support on Unix and Windows systems.

![math-support](https://user-images.githubusercontent.com/52726/173204087-ca836dba-f949-47a8-ac0b-085900eda86a.png)

`mimeTeX` does not depend on a Latex distribution. However, unfortunately it can only write GIF files. 
I made a brief attempt at resurrecting GIFReadWriter to get rid of the ImageMagick dependency but that does not seem to be as easy as I thought it would be (at least for me).

There is probably plenty of room for improvement. Feel free to leave a comment.

andswitch committed:
Support Latex math formulas on Unix

Formulas are enclosed in dollar signs: $x^2+y^2$.

The Morphic renderer converts formulas to images using the external OS
processes `mimetex` and `convert`. The images are then inserted as
regular images.

The Latex renderer renders formulas as inline formulas.

andswitch committed:
Add Plot support for gnuplot on Unix

Plots are inserted by enclosing the gnuplot script between <plot> and </plot>.
They are automatically converted to images and the images are inserted
into the document at parse time.

andswitch committed:
Typo

andswitch committed:
Use transparent and truecolor options for Gnuplot PNGs

mmontone commented:
Hi,

this looks interesting!

Do you think there's a chance of providing math + plot as a separate package?

For example, instead of modifying EruditeMarkupGrammar, subclass it and put it in an Erudite-Math package. Then I can add a way of specifying a custom grammar class for a Book.

Not because I don't like this, but because this relies on external dependencies.

What do you think?

mmontone commented:
For example, you can create your separate subclass of EruditeMarkupGrammar and do:

```
!MathEruditeMarkupGrammar methodsFor: 'as yet unclassified' stamp: 'AS 6/11/2022 19:55:37'!
markup

	^ super / math / plot! !
```

andswitch commented:
Besides the EruditeMarkupGrammar class, there are also changes to the following classes:

1. EruditeMarkupParser (parsing math into DocMath nodes + transforming plots into DocLink images)
2. MorphicEruditeDocRenderer (rendering DocMath nodes as image)
3. LatexEruditeDocRenderer (rendering DocMath nodes as text between dollars)

Can the same trick be used for all these changes?

mmontone commented:
I think so. I think you need a custom parser class and a custom renderer class.

But let me try unhardcoding the parser and renderer classes first so you have something to work with.

andswitch commented:
Great! Also, do you know of a Cuis-compatible GIF reader? That could eliminate the ImageMagick dependency as it's only used for converting GIF to PNG..

mmontone commented:
No. I don't know if there's a GIF reader.

mmontone commented:
I've made some changes to support custom parsers for books (see latest Erudite commit).

Can you try this?: (worked for me, at least partially).

EruditeMathMarkupGrammar is subclass of EruditeMarkupParser. That's odd, but that's the way I found to extend parsers with what Smalltalk provides.

Also observe #parser method on class side of book class.

I've only changed the parsers of books, but not the renderers. That'll probably need more work.

EruditeMath.pck.st:
```
'From Cuis 6.0 [latest update: #5234] on 11 June 2022 at 8:01:09 pm'!
'Description '!
!provides: 'EruditeMath' 1 1!
!requires: 'Erudite' 1 178 nil!
!requires: 'Cryptography-DigitalSignatures' 1 14 nil!
!requires: 'OSProcess' 1 23 nil!
SystemOrganization addCategory: 'EruditeMath'!


!classDefinition: #EruditeMathMarkupGrammar category: 'EruditeMath'!
EruditeMarkupParser subclass: #EruditeMathMarkupGrammar
	instanceVariableNames: 'plot'
	classVariableNames: ''
	poolDictionaries: ''
	category: 'EruditeMath'!
!classDefinition: 'EruditeMathMarkupGrammar class' category: 'EruditeMath'!
EruditeMathMarkupGrammar class
	instanceVariableNames: ''!

!classDefinition: #EruditeMathMarkupParser category: 'EruditeMath'!
EruditeMathMarkupGrammar subclass: #EruditeMathMarkupParser
	instanceVariableNames: ''
	classVariableNames: ''
	poolDictionaries: ''
	category: 'EruditeMath'!
!classDefinition: 'EruditeMathMarkupParser class' category: 'EruditeMath'!
EruditeMathMarkupParser class
	instanceVariableNames: ''!

!classDefinition: #MyMathBook category: 'EruditeMath'!
EruditeBook subclass: #MyMathBook
	instanceVariableNames: ''
	classVariableNames: ''
	poolDictionaries: ''
	category: 'EruditeMath'!
!classDefinition: 'MyMathBook class' category: 'EruditeMath'!
MyMathBook class
	instanceVariableNames: ''!


!EruditeMathMarkupGrammar methodsFor: 'as yet unclassified' stamp: 'MM 6/11/2022 19:37:24'!
markup
	^ super markup / plot! !

!EruditeMathMarkupGrammar methodsFor: 'as yet unclassified' stamp: 'MM 6/11/2022 19:13:00'!
plot

	^ '<plot>' asParser flatten,
		'</plot>' asParser negate plus flatten,
		'</plot>' asParser flatten! !

!EruditeMathMarkupParser methodsFor: 'as yet unclassified' stamp: 'MM 6/11/2022 19:46:35'!
plot
	^ super plot ==> [:res |
		| script plotId pngPath |
		script := res second.
		plotId := (((SecureHashAlgorithm new)
			initialize;
			hashMessage: script)
				printStringBase: 16) asLowercase.
		pngPath := '/tmp/erudite-plot-', plotId, '.png'.

		pngPath asFileEntry exists ifFalse: [
			| gpPath gpScript proc |
			gpScript := 'set term png', String lf,
				'set output "', pngPath, '"', String lf,
				script.
			gpPath := pngPath, '.gp'.

			gpPath asFileEntry fileContents: gpScript.

			proc := ThisOSProcess thisOSProcess command:
				'gnuplot ', gpPath.
			[proc isComplete] whileFalse: [(Delay forMilliseconds: 10) wait]
		].

		DocLink target: pngPath type: #image label: script
	]! !

!MyMathBook methodsFor: 'as yet unclassified' stamp: 'MM 6/11/2022 19:46:10'!
Intro
^(EruditeBookSection basicNew title: 'Intro'; document: ((EruditeDocument contents: '!!!! My math book

<plot>
plot sin(x)
</plot>

') images: ((Dictionary new) add: ('/tmp/erudite-plot-ea659a68332d07df52e242741687764a9abf024d.png'->(EruditeForm fromBase64String:'iVBORw0KGgoAAAANSUhEUgAAAoAAAAHgCAMAAAACDyzWAAADAFBMVEUAAAAAAAD///9/f3/l
AAAVsBoDQ98A/////xTCAHgfHx8/Pz9fX1+fn5+/v7/f398HBwcPDw8XFxcnJycvLy83NzdH
R0dPT09XV1dnZ2dvb293d3eHh4ePj4+Xl5enp6evr6+3t7fHx8fPz8/X19fn5+fv7+/39/cA
AAAAMwAAZgAAmQAAzAAA/wAAADMAMzMAZjMAmTMAzDMA/zMAAGYAM2YAZmYAmWYAzGYA/2YA
AJkAM5kAZpkAmZkAzJkA/5kAAMwAM8wAZswAmcwAzMwA/8wAAP8AM/8AZv8Amf8AzP8A//8z
AAAzMwAzZgAzmQAzzAAz/wAzADMzMzMzZjMzmTMzzDMz/zMzAGYzM2YzZmYzmWYzzGYz/2Yz
AJkzM5kzZpkzmZkzzJkz/5kzAMwzM8wzZswzmcwzzMwz/8wzAP8zM/8zZv8zmf8zzP8z//9m
AABmMwBmZgBmmQBmzABm/wBmADNmMzNmZjNmmTNmzDNm/zNmAGZmM2ZmZmZmmWZmzGZm/2Zm
AJlmM5lmZplmmZlmzJlm/5lmAMxmM8xmZsxmmcxmzMxm/8xmAP9mM/9mZv9mmf9mzP9m//+Z
AACZMwCZZgCZmQCZzACZ/wCZADOZMzOZZjOZmTOZzDOZ/zOZAGaZM2aZZmaZmWaZzGaZ/2aZ
AJmZM5mZZpmZmZmZzJmZ/5mZAMyZM8yZZsyZmcyZzMyZ/8yZAP+ZM/+ZZv+Zmf+ZzP+Z///M
AADMMwDMZgDMmQDMzADM/wDMADPMMzPMZjPMmTPMzDPM/zPMAGbMM2bMZmbMmWbMzGbM/2bM
AJnMM5nMZpnMmZnMzJnM/5nMAMzMM8zMZszMmczMzMzM/8zMAP/MM//MZv/Mmf/MzP/M////
AAD/MwD/ZgD/mQD/zAD//wD/ADP/MzP/ZjP/mTP/zDP//zP/AGb/M2b/Zmb/mWb/zGb//2b/
AJn/M5n/Zpn/mZn/zJn//5n/AMz/M8z/Zsz/mcz/zMz//8z/AP//M///Zv//mf//zP/////A
vQAOAAASBElEQVR4XuXdCZbavBKGYWkd3o/WIU/730I8AI3ABAOSqkpvnXvSyf3T6EvVg83U
luMXv/jFL37xi1/84he/+MUvfvGLX/ziF7/4xS9+8Ytf/OIXv/jFL37Rq/cdvwn8kqo4TnyA
/JKreXZ8gPwSLT5AfvEB8osPkF/84gPkFx/g8v/wi1//q+IAtdw1tl+DC0FDjve1xAwls+ac
yy9BWQDD3i4jAItmzTiX8IvBnD6GGH0Xo16AIRmuaoDFo/qs/r6PmtPHtJ/UB60Aw1PP1AIs
H9Xn9qcAoMwCn3dK9hDYWNRf7yscgOHlH1QDbD0qH6DyoYpEHZPnkt14+92UPSoGYPjvHxUD
lIg63z+Oj3eP6qc+d1f5ALX7E4869scYj7IFPsCTU9UsUC5qXF7ImOJ2Ch7nbvTTsJubfVx/
Mz8eAoPjA/xyqIoBCkb1/TB0C7gV4NgNcT349esjwGm62Pt7PHicLPABnpyqXoFyUYf1QOeG
/Qi4Pt/op8sZePDzPA5P52A+wO+Hygd49HBv7OLlWfD2RHg93u3PiOdxnP+M/idX4AM8CVCr
QMn7yrA87luh3QP08z7H8XKQnNNc4ZPiA7QJsGbUoV+Occ9HwG7cv/KPgNmGqlRgEEw6bLiW
Y1wCsN8e+sW40bt/DJglKh8gH+DfqzB+/nsWfAW4PhFxt6fD98+Cs3SVD9ACwFpR57vXAa8A
12Netz4DHtYPe96/DsgH+NtQVQJUGPXlOyF8gD9OVaNAhQBfvRecp6t8gBb8yUZ98WkYPkDT
UzUflQ+QD7CFqGiA+sbKi8oHaGOofIB2ARqaKjAqH6CRqWrLygcIm2qrUfkAjQyVD9AqQENT
Dc5M1mxd5QPkA+QD5E+VG7V5gMHxAfIB8gFmiKopa76u/uhj3n+IZa/1J1o6PkA+wHoAZz8P
3fXHVPoxDvHpEqmeP1V+1GIAtwNef/mQ2GavH3UBDI4PsF2A+8/ozZ4PEAYwY1d/8rF/WDte
zsHdGPcL2PABlomqJ6sWgPux7/azyr336q6S39JUm4yaEeD2JGTU9SQkOD7AhgEmp+D9qiHz
4w0W2A2HD1B90nNRf7aRPAl5vnQD/wjIj1r2CLj/zPI0uf8dAflTzRZVS9acUTO9EN0ttzNt
jwF7TQCDY04VA/D2VtwKUOFbcXyA6rva9ocR+AD5APkAmVHPZuUDVDJWPkA+QMePmjEDH2Du
VvEB8gHyAdoBePLv8QG2eFjhAzQFUMNY+QD5ABuaKh8gH2CzUYPjA5RpFR8gH2ChVsmP1dB9
JXNX+QD5APkA+QC5UfkA+QD5APFTDQ4btWGA4KnyARoDKD3WNgGe+rt8gMYAthaVD5APkA+Q
PtXg+AD5APkARaK2C5A8VT5AawBlx9oqwDN/mw/QGsDGovIByk81OD5APkA+QD5A6lTRUfkA
+QD5APlTtZJVHcBkp6Q4Kbo8W7tTbStqxp2ShrEfZjUXKbcz1eD4AL+tZKek7WuMfIB8gLUA
pjslPZ9+BQEamio7atZtGnr/dIloPsACKwc+wOtDwL/j4PJlnIdZy26ZfIBAgP3y63a1clfh
6lhvig/QSNSffBzslPS8UY3MTknsqRqJmnenpH27Vi07JdmZanD8I+DXNfbutlPS/rXTsV+w
oanCo/7mI9kpKfo+ankS0vZUpbLqA5julKTorTg+QAjA98UHiAb49nv4AFucKh+gMMDg+AD5
APkA+VH5ANucqkxWPkA+QD5APkA+wKzfxQfIB8gHqKNTDU2VD5APsD2AZbrKB9jiVPkA7xYI
/Knyo/IBNjZVF1qJygfIB8gHqKRT7UyVD5APsDmAhbrKB8gHyAfIBciPygfIB1h6xSALsH6v
+FPlA5QEGBx+qnyAfICtASzVVT5APkA+QD5APkA+wJamqivqf7+XD5APsHWAtXsVnJmx8qPy
AfIB8gHq6RQfYPWu8gHyAfIBUgFSov7vuzNdH/Ba48gHyI9aD2CyVZdbr1NpHGDdrHyAP18j
+n6rrm2nBj5APsB6ANOr5Ds3dd0RQENT5Uc1BTDZJ2S7Qj4fID9qRYDJTknbV3mAwfGnaqmr
GQFOveMD5Ef98PvznYLncTgEuFQbreIDzP39ebfq6v1enekjIB+gpSNgslXXsFY3Do4PkA+w
FsBkq67LE2E+QH7UagDTrbpeAuRPlR+1EMD3xQdYbCU+QIUADU2VH5UP8Fp3xY9aZKWXt8AH
KDZWPkA+QD5APkA+QHbUSts08KfKj8oHqGuqwfEB8gHyAfIBUqfKj8oHyAdYb53AB6hsrHyA
fIDWAbYQtdZumfyplliFD5APkA+QD5APkA9QCUBDU+UD5APkA1QQlQ/Q8lTrZC0alQ+wxany
AT4vwJ8qPyofoJ6pBscHyAfIB8gHSJ0qHyAfIB+ghqh8gKanWiNrIwDtt4oPsMTt8AG2OFU+
QD5AflQ+QMVTDY4PMAvAZKekbvlDJwjQ0FT5APMATHZK6v0cu8cr9PIB8qMWBJjslPSwbRIf
YJUVAhng405J7uASqb6ZVvEBqgP4sFPSWutmIXyAfICnbynjRjXbAdAPDQDkR7UKsPOzEwMY
HH+qMICPp+D+/mR8XeC2Gw4fIP++8spGnich3YE//hGQH7XcETDZKWmBGB0fID9qTYDJTklj
H9d6vUDgT7XA7QcywPudkuJls8KBD5AP8Pxt1fswAh8gH2DTAIPjT5UPkA+QD5APkD9VQ13l
A2xxqnyAfID8qPoAlu0VdaqGunp0a3yAfIB8gArHEPgA+QD5APkA+QDZAMt3lQ/wuH64ufeV
98ZLRg3Fb44PsMWp8gHyAfKjnru9qgBL9oo/VT7AlgDyo/IByg6BD5APkA9QOmqFrvIB8gHy
AfIB8gHyAbY0Vb23HPgA+QD5AG22ig/QPkBDU+VH5QOUHQEfIB8gH6Bs1Bpd5QPkA+QDRAFk
R326TT5APkA+QBut4gNUDzDZN4kPkA+wMsBk3yQ+QH7U2gAPt0ry/Knyo9YBeLBvUi2AwfGn
arSrGQEe7JvEB8iPWg/g875JfID8qO9ulQ+wBSt8gP85Bae74fAB8gG+ssF/EsIHaPoImO6b
xAfIj1ob4IkXovlT5UctB/DEW3H4qRp6cdEgwKPy/Knyo/IB8gGqifpwu3yAdW+XD5APkA+Q
D5APkB+VD5APEAmQPVVDP2rHB8gHyAdoAyD/J3j5APkA+VH5AA2MlQ/w8Zb5ANsAaDYqHyAf
IAwgeaqGLjzNB8gHyAfIB8gHyAfY0lT5APkA+QB1ReUDrHnb/Kh8gHyAuqIKAORP9YM6X3yA
x8UHWCcpHyAfIB8gEqChqfIB8gHyAfIB8gEayFqtq3yAfIA4gIZaxY9aOiofIB8gHyBkqmWT
QgEmV2Trlj90lQEamiofYH6AyTUpez/Hznd8gHyA1QAmmyOd2SmJD9BE1npd/Q3gwXXJu5EP
kA+wFsCDnRmm/j1Am63iA9QH8Hlvmm4j2Q5AftTCUTMD7Pzs+AD5USsAjN77/vEU3D9u0vBi
N5xgplPMqJUA/rRT0hBjHB6ehHTP/kofAfkA0afgdHOk+cAfHyA/akmAtxeiO79qjGvxAQpF
zbeCHYC3t+IWgNHvNbwHaLJVfIAaAb4tPkA+wP+twAfY4lT5AGsBLN8pPkA+QD7AyivU7Cof
YItT5QPkA+RHVQ0w17+RP1XrXeUD5AMUjcoH2OJU+QArAazRKT5APkA+wKprVO0qHyAfIB8g
ASA/Kh8gH6CiqNdVpADm+Vfyp2q2q3yALU+VD7AOwDqd4gPkA+QDrLhK3a7yAfIB8gHyAfIB
8gG2NFULUa8/nM4HWGUdPkBtAA1NlR+VD5A/VT5AfQBrdYoFsHJX+QD5APkA2wfIj8oHyAeo
LOq+Eh9glZX4APkA+QCdaxBgslPSUuN4FqChqfKjqgWY7JTk1ssE8gHyo1YE+LA50uD7egDr
dYoEsHZX8+6UNHUdHyA/aj2AD9s0LPr4APlRP1or50Y169c2Af66Fh9gFYDrNnF8gHyAdQA+
75Q0j8MRwJe74fAB4gHm3Smpv2zT0J08AhqaKj+q0lNwslPSsFY3Do4PUDApDGCyU9LliXAl
gIamyo9aDOD9Tkl8gPyo9QG+rTYA/rYaH+Dr1fgA+QD5APkAuVElAfKnyo/KB1hhtbpJ+QD5
APkAAQANTZUflQ+QD1Bl1GU9PsDy6/GjagXInyo/Kh9g8fVqJ+UD5APkA2weoKGp8gHyAfIB
8gHyARrIKtBVWYCGWsWPWiYqHyAfIB9g01Otn5QPsDxAQ1PlA+QD5APkA+QDNJBVoqvCAC21
ig+QD5A/VT5APkB+1Io+lAKU6FT7UfkA+QD5APkAW4j63aoiUaUBGmoVHyAfIB8gHyAfIB8g
HqDMUA1NFQQw2aorTn7s+AD595V6AJOtuoaxH+ZHgXyA/KgFASZbdW1fY/xwgcCfqo6sFgGm
W3U9nX75APkAywJM9gkZ/Nz77bL5fIB8gHUAJjslDcvTkWH2XXGAUkM1dF+BAlyPfp3nAzQJ
UChpxq26llPw7ax8t8C77ZiaBij0aPWbhT/9jgzl827V5baz7xPAdzf0+T9coFXmon6+slDU
jFt17V8fd6rxzbSKD1AhwGSrrrickz9/EmIIID+qOoDpVl3fvBX3+b9cbqh8gPoAZliAD1BD
VKmkfIB8gHyAzQLkR7UA8NN/O3+qJdbmA+RPlQ/QBEDJofIB8gHyARZZWywpHyAfIB9gqwD5
UW0A/OxfLzpVQ1H5APkA+VEbBCg71FYByiXlA+QD5ANsFCA/Kh+g6Fj5APkA+QD5ANuaKj8q
HKD0UNsEKJiUD7DFqfIB8gHyo/IBCo+VD5APkA+QD9A3NtXTCeSTmuiqMYCGpsqPygfIB8gH
2B7Asxn4APkA6QAtROUDLDRVDUn5AM8vEPgA+QD1AzQ0VX7UOgCTnZK68ZvLs1mb6rkUfIBV
ACY7JfVjHOLnF6jkAyRHzblT0mavH79boLGp6kjaPMCDi5TzAfIB1gOYbNOwPASMy+/ncgAN
TZUPsArAZKOa5ejnvf/ySYilqZ7JwY8qAHB7EjJ2Xy7Anyof4Gdn3+OdkubHrbrO7obT1FS1
JNXd1cw7Je0OP94pqcmp8gHWOQUnOyW9OALyAfKjlgOY7JQ0bY8B+28XCPypSmS1DTDdKemH
t+JMAXybRE/S5gFmXKChqfIB8gHyAdqIqgjgu17wp1oiKx8gHyAfoAmAmobaTtR3/7l47T4K
/u+DBcIP/7Vy8aMK+Ci/QOBPtXZWPkA+QD5AEwB1DdXQfUV5V/kAW5wqH+CXCwT+VOtm5QO0
CpAfFQZQ21DbAKggKR9gi1PlA/x2gWBmqvyoKID6hsoHyAfIn2oDXVUG8FVPFAI0FJUP8Nde
afTHv6/wAWqcKj8qH6DoWPkAbQM87IpOf3yAfID6xsqPygcoOlY+QD5AybFqTao3qjqA/Kny
AfIBfp2VH9U+QMtT1ZtUbVQDABVP9TEbP2oLAO1OVXNSCMDedxkWCGamyo+qCmAcp+wAdQ/V
atRGAc6zywLwrj3Kh5oEDPyo8o8B8wD8a1DQ/mD30/uK1wAwiOawBDDof7b1YVSv4BAYHB/g
216FkPdMUQ5g+OiuIjr4PWtwfIDnDaoHeAlqAuBGMLgGAW77Jh0D5Be//ld5AG77Jh0C5Be/
qh5R+QD5JVbLcdB3MfIbwS+Zmvaz+sDvBL/4xS9+8Ytf/OIXv/jFL37xi1/84he/+IWp6+fz
r7uqC9b+CmUv3hIFrVDTjtI8bp/Pn/08dD7KdryP13erRf0paIWSdpTncft8/rj+2k+iHR91
vEWtoRVK2lGFx7bCsOmePR+gjlboaUd5HtsKcXtvOMqeeHw/efmuq2iFnnaU57GtsOMeZLs+
TnPsxD8qpqIVetpRnocegFt1ng/wrsTbUYqHSz+fL3zeiX+vN+xJBEvNKVhHO4rxSD+fv+Pe
oQvU8Pd6wyzdceFWPNSsAmDRnmwruHE9AE2yrz1sre7FJ6+hFXraUZjH7fP5Cl59Hfzyyqv8
o26n5IVoDe0oz+Pv8/kK3n+Kk473wJS8FaegHap48Itf/OIXv/jFL37xi1/84he/+MUvfvGL
X/ziF7/4xS9+8Ytf/OIXv/jFL37xi1/84lfz9Q/7lc90bR/13AAAAABJRU5ErkJggg==')); yourself); yourself); subsections: (OrderedCollection new yourself); yourself)! !

!MyMathBook methodsFor: 'as yet unclassified' stamp: 'MM 6/11/2022 19:46:10'!
initialize
    super initialize.
    title _ 'MyMathBook'.
        self addSection: self Intro.
! !

!MyMathBook class methodsFor: 'as yet unclassified' stamp: 'MM 6/11/2022 19:33:43'!
parser
	^ EruditeMathMarkupParser new! !
```

mmontone commented:
Now this got me thinking. Because I think it would be cool to add PlantUML extension, using your same approach.

But then this way of extending the parser by subclassing doesn't work. It is too rigid and too limited. Say I would like to have Math and PlantUML extensions, how to do it? It is impossible with subclassing.

I think it may be possible to do it in another way, by allowing the extension of particular parsing rules. And then have:

```Smalltalk
MyBook class>>parser

    ^ EruditeMarkupParser with: EruditeMathParser with: EruditePlantUMLParser
```

I'll see if I find a way of doing that.

andswitch commented:
> Say I would like to have Math and PlantUML extensions, how to do it? It is impossible with subclassing.

Personally, I wouldn't try to make that work without laying down ground rules for extending the Erudite language. Otherwise you might end up with a confusing framework. At the end of the day, I would want that all the extensions integrate nicely. How to enforce that? Both for the extension author and the framework author.

My gut feeling is the ability to override grammar, parser and renderers per book is good enough and also more simple and flexible.

mmontone commented:
> At the end of the day, I would want that all the extensions integrate nicely. How to enforce that? Both for the extension author and the framework author.

So do I. But how to make extensions integrate nicely using subclasses?

andswitch commented:
The new parsers / renderers don't need to be subclasses of existing parsers / renderers. They can reuse existing ones however they want. Maybe composition instead of inheritance.

If there is only one object for the parser, one for the morphic renderer and one for the latex renderer per book, there is only one language per book, so the author of the extended language has to make sure everything they want to reuse works together nicely.

If somebody wants to combine two extensions, it's up to them to create one extension that combines both of them.

You could say this isn't really solving the problem because now the burden falls squarely on the extension implementor. I'm not sure what else you could do that doesn't put unforeseen limitations on extensions. Or maybe I'm just seeing problems that aren't real.

mmontone commented:
> If somebody wants to combine two extensions, it's up to them to create one extension that combines both of them.
> 
> You could say this isn't really solving the problem because now the burden falls squarely on the extension implementor.

I would say that the burden falls on the user too, because he may not be allowed to use several extensions at once.
I would like to at least explore the possibility of providing a way for users to combine extensions, and implementors to implement extensions in a way that the extensions can be combined.

mmontone commented:
I've added an EruditeMath package here: https://github.com/Cuis-Smalltalk/Erudite/blob/master/EruditeMath.pck.st

so we can work on it freely using git, is that ok with you?

For the renderer, it is easier to extend, because there's no need to overwrite any method, like in parser; and so Cuis package extensions suffice. Still ... I'll consider implementing custom renderers per book.

Please have a look and let me know what you think.

mmontone commented:
Added a #morphicRendererClass to EruditeBook class that can be customized for a book if desired.

andswitch commented:
I noticed that the EruditeToolsExtensions uses the parser and morphic renderer outside of a book context. So, now I'm not sure that the customization should be per book.. See `BrowserWindow>>buildMorphicCommentPane`.

andswitch committed:
Merge branch 'master' into math

andswitch committed:
MyMathBook -> EruditeMathManual

andswitch committed:
Add System Requirements section

andswitch commented:
> Please have a look and let me know what you think.

The MyMathBook opens no problem! I renamed it to EruditeMathManual and added some sections to describe the syntax and how to create a new math book.

I'm already quite happy that this works!

andswitch committed:
Small typos

andswitch committed:
Revert "Small typos"

This reverts commit 67957dd0c18e56d444061eb87600c03c224b7b07.

andswitch committed:
Repeat small typos

mmontone commented:
Very good!

El dom., 12 jun. 2022 15:03, andswitch ***@***.***> escribió:

> Please have a look and let me know what you think.
>
> The MyMathBook opens no problem! I renamed it to EruditeMathManual and
> added some sections to describe the syntax and how to create a new math
> book.
>
> I'm already quite happy that this works!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Cuis-Smalltalk/Erudite/pull/5#issuecomment-1153248831>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADKPDSBWKKVK7XNX4KDZTDVOYQ6DANCNFSM5YQR6GNA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

andswitch committed:
Add transparent truecolor to Gnuplots

andswitch committed:
Convert formulas and plots to dark-mode compatible colors if the current theme is dark

andswitch commented:
Added a mechanism that converts math and plots to light colors if the current theme is a "dark mode" theme.

mmontone commented:
I'm using this code in some PlantUML code I have:

```Smalltalk
OSProcess waitForCommand: command.
```

Doesn't that work instead of what you have now?

mmontone commented:
I'm also using this for generating random filenames:

```Smalltalk
getTemporalFileName: extension

	^ '/tmp/plantuml-', 'abcdefghijk' shuffled, '.', extension
```

Not sure how robust that is, but you can get rid of the Crypto dependency if you decide in favor of something like this.

Yes, not very robust, but still ...

mmontone commented:
> ```smalltalk
> 'abcdefghijk' shuffled
> ```

I've changed by this now, a bit more robust:

```Smalltalk
getTemporalFileName
	
	|randomStr|
	
	randomStr _ String streamContents: [:s |
				15 timesRepeat: [s nextPut: ($a asInteger + ($z asInteger - $a asInteger) atRandom) asCharacter]].
	^ '/tmp/plantuml-', randomStr
```

andswitch commented:
It would be nice to get rid of the crypto dependency, but I wanted to have deterministic filenames so that you can safely skip generating images for formulas/plots when those images already exist.

mmontone commented:
Ah. I see. Good point.

mmontone commented:
![EruditeUML](https://user-images.githubusercontent.com/436110/173259338-8bad89f3-c277-449c-829e-5d24df4f6b69.png)

I've added an UML extension using PlantUML inspired by your work :)

It is a pity we don't have a way of combining the extensions. At least not yet ...

andswitch commented:
How do you feel about bundling all the features that require external dependencies in one `EruditeExtra` package or something similar?

mmontone commented:
> How do you feel about bundling all the features that require external dependencies in one package and name it something like EruditeExtra?

Yes. That crossed through my mind too.

I think it could be a good idea ..

Perhaps rename the grammar classes to something like EruditeExtraMarkupGrammar, and include plot + math + uml there.

andswitch committed:
Use OSProcess waitForCommand

andswitch committed:
Reorganize dark-mode logic to use package protocols

andswitch committed:
Remove Unix requirement

andswitch commented:
I removed the requirement to have a Unix by placing the temporary files in the current working directory. Maybe it's not so nice, but i.e. asciidoctor-mathematical does something similar. At least it is OS-agnostic.

/edit: but I guess it would be nice to put the files in the OS's temporary directory.

mmontone commented:
> I removed the requirement to have a Unix by placing the temporary files in the current working directory. Maybe it's not so nice, but i.e. asciidoctor-mathematical does something similar. At least it is OS-agnostic.
> 
> /edit: but I guess it would be nice to put the files in the OS's temporary directory.

I think that's a reasonable trade-off. Is it not possible to implement a #getTemporaryDirectory method that returns the temporary DirectoryEntry according to the running platform?

I looked in OSProcess package, but haven't found anything yet..

mmontone commented:
> I removed the requirement to have a Unix by placing the temporary files in the current working directory. Maybe it's not so nice, but i.e. asciidoctor-mathematical does something similar. At least it is OS-agnostic.

Perhaps a bit better would be to use current directory as base, but put files in a 'tmp' subdirectory.

andswitch committed:
Use OS-temp directory on Windows and Unix

andswitch commented:
Used `FileIOAccessor` to check if we're on Windows or Unix and then act accordingly. What do you think?

mmontone commented:
> Used `FileIOAccessor` to check if we're on Windows or Unix and then act accordingly. What do you think?

I think it's good enough in practical terms.

I was trying with OSProcess and environment variables (I see that's how other systems do it), like:

```Smalltalk
|thisOS|

thisOS _ ThisOSProcess thisOSProcess.

thisOS class caseOf: {
	[UnixProcess] -> [((thisOS environmentAt: 'TMPDIR')
					ifNil: ['/tmp']) asDirectoryEntry].
	[WindowsProcess] -> [(thisOS environmentAt: 'TEMP') asDirectoryEntry].
	[MacProcess] -> [((thisOS environmentAt: 'TMPDIR')
					ifNil: ['/tmp']) asDirectoryEntry].
	}
```
But this is untested, so ...

andswitch committed:
Use environment variables for determining temp dir

andswitch commented:
Works on my machine. :)

andswitch commented:
The current approach does have one downside though:

Suppose person A wants to share a math book that contains some formulas/plots with person B and

Person B:
1. Does not have the external programs installed
2. Is on a different OS

Because the temporary-files directory is different per OS, the generated image path for each formula/path for person A is different than those for person B.
The images for the plots/formulas will not be found in person B's document's image dictionary because the paths are different.
And since person B does not have the external tools installed, the images can also not be generated.
So person B will not be able to see the formulas/plots.

Storing the images in the current directory would solve this problem, as the image paths would be the same across OS-es so the images can always be found in the dictionary. Maybe it's worth it after all?

mmontone commented:
Images (like plots) are serialized with the books. And for rendering, they are fetched from the "image cache" (see 'images' instance variable in EruditeDocument). The 'key' of that image Dictionary is the image path, and the 'value' is the actual Form for the image. For rendering a document, when a DocLink with type #image is found, the renderer tries to find the image in the cache, using the DocLink path as key.

So, my guess is that the book transfer will work, but not when on different platforms, because the temporary directory name is used as part of the image key.

andswitch committed:
Let Latex renderer use custom Markup parser

mmontone commented:
> Because the temporary-files directory is different per OS, the generated image path for each formula/path for person A is different than those for person B.
> The images for the plots/formulas will not be found in person B's document's image dictionary because the paths are different.

Ah, I see you are considering what I tried to explain. Yes ...

andswitch committed:
Make plot a proper syntax node.

Render plot differently in Morphic and Latex

andswitch commented:
I think it would be helpful to automatically remove unused images from the document image dictionary when the user saves the book in the editor. Or does that already happen somehow?

mmontone commented:
You are right. I didn't pay attention to that. I'll have a look.

andswitch committed:
Make math book readable across OS-es and themes

Store generated formulas/plots in the document image dictionary with
keys that are the same for every OS.

Images are still generated in OS-temporary directories.

Generate dark-mode and light-mode images at the same time, so that the
book can be shared and read under dark and light themes.

andswitch commented:
> > Because the temporary-files directory is different per OS, the generated image path for each formula/path for person A is different than those for person B.
> > The images for the plots/formulas will not be found in person B's document's image dictionary because the paths are different.
> 
> Ah, I see you are considering what I tried to explain. Yes ...

I managed to find a solution that solves the problem while at the same time still generating images in the OS's temporary directory: instead of using the filepath as key in the image dictionary, formulas/plots use a different key that is OS independent.

mmontone commented:
> I managed to find a solution that solves the problem while at the same time still generating images in the OS's temporary directory: instead of using the filepath as key in the image dictionary, formulas/plots use a different key that is OS independent.

Well done. I thought about doing that, but didn't know exactly how.

andswitch committed:
Regenerate math manual images

mmontone commented:
> I think it would be helpful to automatically remove unused images from the document image dictionary when the user saves the book in the editor. Or does that already happen somehow?

I've had a look but I find this tricky to implement. I tried resetting the images Dictionary and re-render the book before storing, but then you lose the ability to edit books without having images on disk, which is possible atm, something very useful to have IMO.

mmontone commented:
I'm not very happy with the book storage. It is reliant on the renderer, and it should not. Also, the images cache and transfer is a bit of a hack. I think I'd prefer storage to only rely on the model (the parsed contents document). I tried serializing the Document object but the result was too ugly to my eyes.

andswitch commented:
> I'm not very happy with the book storage. It is reliant on the renderer, and it should not. Also, the images cache and transfer is a bit of a hack. I think I'd prefer storage to only rely on the model (the parsed contents document). I tried serializing the Document object but the result was too ugly to my eyes.

What if the image cache is moved to the model? That way you can still store generated images and let the renderer decide to use them or not.

mmontone commented:
It is in the model already. Instance variable of EruditeDocument class. But images are not fetched and stored in there until they are rendered. The document model only contains references to image files. Perhaps instead of storing that image cache, I should go through each of the DocNode objects and ask them if they have an image to store.

andswitch commented:
I see. I didn't realize that. The only reason I chose to generate the images in the renderer is because the parser doesn't have access to the document (or does it??), so I thought you can't store images during parsing.

mmontone commented:
I'm having an idea of having DocNodes expose binary data that need to be transferred. For example, file contents. And be neutral about format, so they can be images or other things, unlike now that they are images as Form objects. Then those are serialized. But I will have to work a bit on that.

andswitch committed:
Fix DoesNotUnderstand exception in enableEditing:

mmontone commented:
> I'm having an idea of having DocNodes expose binary data that need to be transferred. For example, file contents. And be neutral about format, so they can be images or other things, unlike now that they are images as Form objects. Then those are serialized. But I will have to work a bit on that.

I think I may have come up with a good implementation of this.
For DocNodes that use and need data serialized, they should implement #dataKey and #data methods.
See DocLink class for an example.
Also look at ImageDocLinkRenderer for how to access the node's data from the renderer.

Can you try it and let me know how it goes?

andswitch commented:
> Can you try it and let me know how it goes?

Will do, small request in the meantime: I made two one-line bug fixes in Erudite.pck.st in this branch, it concerns these lines:

https://github.com/Cuis-Smalltalk/Erudite/pull/5/commits/866e6fe06f2a849780c8fb1751f4a512c048ca32#diff-03dde013b47bf702197b1f8e1af7261766e77cbd4b05191ee05ea31f3d14ab02R13099

https://github.com/Cuis-Smalltalk/Erudite/pull/5/commits/664fbf040fca5ea89e212b3b9a00e0728b300ddf#diff-03dde013b47bf702197b1f8e1af7261766e77cbd4b05191ee05ea31f3d14ab02R14155

Could you have a look and maybe take them along (if you didn't do it already)?

andswitch committed:
Merge branch 'master' into math

andswitch commented:
Is there a way to get the current document from a `visit..:` method in the Latex renderer? It seems it doesn't have this instance variable.

andswitch committed:
Update to master

andswitch commented:
> I think I may have come up with a good implementation of this. For DocNodes that use and need data serialized, they should implement #dataKey and #data methods. See DocLink class for an example. Also look at ImageDocLinkRenderer for how to access the node's data from the renderer.
> 
> Can you try it and let me know how it goes?

This works well! I was able to get the morphic renderer working and it also looks like unused formulas/plots are automatically removed as well. :)

Only issue is that the Latex document also depends on some gnuplot-generated Latex code for plots. Currently, that code generation is still triggered from the Latex renderer because I couldn't figure out how to access the document data from there.

mmontone commented:
I'm glad the change in the model seems to have worked.

I'll have a look at the LatexRenderer.

mmontone commented:
I applied your changes. Also created a common DocumentRenderer class. And split the Latex renderer into a BookRenderer and a DocumentRenderer.

You should extend the DocumentRenderer for your usecase.

mmontone commented:
Extend LatexEruditeDocRenderer to be more precise ...

andswitch commented:
I think I'm almost there with the Latex renderer. There are apparently many different ways to include a gnuplot in a Latex document. :)

I'm considering to use the `gnuplottex` package. Is it possible to add extra `\usepackage` clauses to the Latex document?

andswitch committed:
Merge branch 'master' into math

andswitch commented:
A small but annoying detail is that the `gnuplottex` package needs to be imported *before* the `minted` package, otherwise Latex will spit out errors.

mmontone commented:
I've added before and after preamble extensions to the Latex renderer.

So, for rendering, use #beforePreamble: aString

```Smalltalk
(LatexEruditeBookRenderer on: MyManual new)
	texFilePath: '/home/marian/src/Cuis/Erudite/CuisManual.tex';
        beforePreamble: '\usepackage{mypackage}';
	render.
```

or for your use case, create/extend a method that uses #beforePreamble: so that you can call:

```Smalltalk
(LatexEruditeBookRenderer on: MyManual new)
	texFilePath: '/home/marian/src/Cuis/Erudite/CuisManual.tex';
        useMathPackages;
	render.
```

Let me know if this works for you, and if you have a better idea on how to do this.

mmontone commented:
Note that I have not worked on the Latex renderer very much, and haven't maintained it in good shape; I do not test it often, I mostly work with the Morphic renderer. So if you see errors, that's expected.

andswitch committed:
Merge branch 'master' of github:andswitch/Erudite into math

andswitch commented:
I also don't intend to use the Latex renderer actually, but I thought it would be nice to make it functional at least. On my machine I can succesfully generate a PDF if `\usepackage{gnuplottex}` is in the right place.

> Let me know if this works for you, and if you have a better idea on how to do this.

I think this will work. The only thing missing (or I'm not seeing it) is that I don't know where to put this code without changing the `Erudite` package. What if there is a `latexRendererClass` method in `EruditeBook` that lets you override the Latex renderer? There is already a `morphicRendererClass` method there. Then `EruditeBookEditorMorph>>exportToLatex` can call this method like so:

 ```smalltalk
(model latexRendererClass on: model)
    texFilePath: '...';
    render.
```

Then I can create a subclass in `EruditeMath` and call `#beforePreamble:` in that class's `on:` method.

mmontone commented:
Ok. I've added the method.

mmontone commented:
> ```smalltalk
> (LatexEruditeBookRenderer on: MyManual new)
> 	texFilePath: '/home/marian/src/Cuis/Erudite/CuisManual.tex';
>         useMathPackages;
> 	render.
> ```

Remember that you could just use Cuis package extension mechanisms instead of subclasses, like in the code above. I think you are already using that for your #visit ... methods in MorphicEruditeBookRenderer.

mmontone commented:
> Then I can create a subclass in `EruditeMath` and call `#beforePreamble:` in that class's `on:` method.

But this also sounds like a nice idea, so ...

andswitch commented:
> > ```smalltalk
> > (LatexEruditeBookRenderer on: MyManual new)
> > 	texFilePath: '/home/marian/src/Cuis/Erudite/CuisManual.tex';
> >         useMathPackages;
> > 	render.
> > ```
> 
> Remember that you could just use Cuis package extension mechanisms instead of subclasses, like in the code above. I think you are already using that for your #visit ... methods in MorphicEruditeBookRenderer.

Yes, I did try creating an `EruditeBookEditorMorph>>exportToLatex` method in `*EruditeMath` protocol, but it seems to also remove the original method from the `Erudite` package.

andswitch committed:
Merge branch 'master' of github:andswitch/Erudite into math

mmontone commented:
Ah. I understand what you are trying to do. Export from the reading/editing tool.

Yes, methods get overwritten by extensions.

andswitch committed:
Windows fixes, Latex fixes

andswitch commented:
So, I think this PR is almost functional now. I took the liberty to make some minor changes in Erudite.pck.st for the Latex renderer.

I'd say the only thing still really missing is an easy way to create EruditeMath books. The manual describes the steps now, but I feel like there should just be a button or something that lets you create a new EruditeMath book. Do you agree with that and do you have any idea on how I should do that?

mmontone commented:
We did not have Erudite extensions before this, so the UI is not prepared to deal with them yet. Let me give it a think. If you have any idea on how to do that, let me know. Only thing is, I'm not sure the way we've extended Erudite so far is in its final form; would be easier to think of UI if we had a more solid base IMO.

mmontone commented:
I think I'm going to merge this. How do you prefer I merge, keep your commit logs or squash them?

mmontone commented:
Also, may I announce this development to the Cuis mailing list, once it is merged? (or you do it, there's no problem...)

mmontone commented:
If you are curious, you can have a look at some experimental stuff I've started. Erudite extensions via MethodAdvises.

MethodAdvises implement advises at method level, like available in Emacs: https://www.gnu.org/software/emacs/manual/html_node/elisp/Advising-Functions.html

With that it is possible to extend Erudite grammar, and combine any number of extensions all at the same time.

Like this:

```Smalltalk
UMLExtension>>installGrammarExtension
	EruditeMarkupGrammar around: #markup
		do: [:nextMethod |	nextMethod value / self umlParser]
```

I know wrapping methods like this has its downsides, but it allows combination in other ways not possible.

https://github.com/Cuis-Smalltalk/Erudite/tree/master/experimental

andswitch committed:
Make it easier to create new EruditeMath books

andswitch commented:
In the last commit I added a way for the user to easily create a new math book. There is a method in the math manual that the user can execute from the manual (Erudite literate programming ftw!)

> I think I'm going to merge this. How do you prefer I merge, keep your commit logs or squash them?

Nice! I also think it's ready to merge now. I don't have a preference for the commit logs, I'm fine with whatever you prefer.

> Also, may I announce this development to the Cuis mailing list, once it is merged?

Feel free to do that.

One more thing I stumbled upon that you should probably know: if you rename a section, a new method is added in the book's class but it seems the old method is not removed.

andswitch committed:
Update Erudite version dependency

andswitch committed:
Make Miktex integration slightly easier

mmontone commented:
@andswitch Thanks for your work here! Cheers

andswitch commented:
My pleasure! I may have a look later to see if it can be based on MethodAdvises. It would be cool to be able to activate both Math and UML for the same book.
